pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+import com.google.common.collect.Sets;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+
+/**
+ * This is the entry point of the Index SPI.
+ *
+ * Ideally, if we used some kind of injection system, this class should be injected into a Pinot context all classes can
+ * receive when they are built. Given that Pinot doesn't have that, we have to relay on static fields.
+ *
+ * By default, this class will be initialized by reading all ServiceLoader SPI services that implement
+ * {@link IndexPlugin}, adding all the {@link IndexType} that can be found in that way.
+ *
+ * In case we want to change the instance to be used at runtime, the method {@link #setInstance(IndexService)} can be
+ * called.
+ */
+public class IndexService {
+
+  private static volatile IndexService _instance = fromServiceLoader();
+
+  private final Set<IndexType<?, ?, ?>> _allIndexes;
+
+  public IndexService(Set<IndexPlugin<?>> allPlugins) {
```

Comment:
I am assuming the `IndexService` will always be created using the serviceLoader. Thus, this constructor can actually be `private` unless there is a specific reason / use-case to keep this at public scope

Reply:
I feel the ctor should still be private. To solve the problem for test and production customization, the factory method should probably take some configuration. 

Otherwise, some code outside index service needs to handle the customized logic. If we take a configuration, by default, it loads from service loader. Otherwise, it builds based on configuration.

Reply:
> Otherwise, some code outside index service needs to handle the customized logic. If we take a configuration, by default, it loads from service loader. Otherwise, it builds based on configuration.

I don't get it. That is precisely the idea. The IndexService is config agnostic. Someone else (usually Pinot startup) is the one that configures it based on the PinotConfiguration received.

Reply:
@gortiz For special distributions use-case that you have explained above (which doesn't yet exist btw), it is still a hassle for the caller to load the filtered list of plugins. For example, in order to disable one index plugin, say bloom filter, the caller will be forced to enumerate and load all other `IndexPlugin` (by service loader or other means) and then, invoke ctor.

```
Set<IndexPlugin<?>> myPlugins = new HashSet<>();
for (IndexPlugin indexPlugin : ServiceLoader.load(IndexPlugin.class)) {
    if (indexPlugin.getType().equals("bloom_filter")) {
      continue;
    } 
    pluginList.add(indexPlugin);
}
```

We can provide more helper util methods for discovery and loading `IndexPlugin`. For example:
```
public static Set<IndexPlugin> loadWithServiceLoader() {
...
}
```

Then, the only entry point for creating this IndexService instance would be via the ctor. Basically, leave it entirely up-to the caller to decide the list of IndexPlugin and only provide helper utils to discover the standard index plugins etc. 
This is an option. 

Otherwise, since currently there is no active use of the public ctor, it seems odd to keep it public when there is a static method to instantiate.

Reply:
I really don't know why do you feel so uncomfortable about having a public constructor. We have tons of classes with public constructors without being a problem. I think it will be very useful for several cases we don't have right now and I would even create a new one that directly receives the set of indexes instead of set of plugins in order to make your case easier to implement. 

But given that we are not using it right now and it is clear some of you don't like the idea of having it, I'm going to make it private. In case we need it in future, we can add just make it public or create a static factory as suggested by @61yao.